### PR TITLE
R4R: support htlt transactions for binance chain

### PR DESCRIPTION
**Pull request by HaoyangLiu**
## Description

Support htlt([Hash Timer Locked Transfer](https://github.com/binance-chain/BEPs/blob/master/BEP3.md)) transactions on Binance Chain.

## Testing instructions
Execute this command to run tests
```
bash tools/tests
```

## Types of changes

New feature

## Checklist

- [ ] Prefix PR title with `[WIP]` if necessary.
- [ ] Add tests to cover changes as needed.
- [ ] Update documentation as needed.

### CODE REVIEW

**Comment by hewigovens** [comment]
Can you add some Swift / Kotlin tests to demonstrate how to use these new orders?

**Comment by HaoyangLiu** [comment]
I'm working on that @hewigovens

**Comment by HaoyangLiu** [comment]
@hewigovens 
I have added some tests for the new features. But the tests are failed for some unknown reason. This is the error message.
```
unknown file: Failure
C++ exception with description "[json.exception.type_error.316] invalid UTF-8 byte at index 1: 0xFF" thrown in the test body.
[  FAILED  ] BinanceSigner.BuildHTLT (12 ms)
[ RUN      ] BinanceSigner.BuildDepositHTLT
unknown file: Failure
C++ exception with description "[json.exception.type_error.316] invalid UTF-8 byte at index 0: 0xFC" thrown in the test body.
[  FAILED  ] BinanceSigner.BuildDepositHTLT (6 ms)
[ RUN      ] BinanceSigner.BuildClaimHTLT
unknown file: Failure
C++ exception with description "[json.exception.type_error.316] invalid UTF-8 byte at index 5: 0xAB" thrown in the test body.
[  FAILED  ] BinanceSigner.BuildClaimHTLT (5 ms)
[ RUN      ] BinanceSigner.BuildRefundHTLT
unknown file: Failure
C++ exception with description "[json.exception.type_error.316] invalid UTF-8 byte at index 0: 0xFC" thrown in the test body.
[  FAILED  ] BinanceSigner.BuildRefundHTLT (6 ms)
```
Do you have encounter similar problems? And could you give me some suggestions?

**Comment by hewigovens** [comment]
can you check `Binance::Signer::build()`? I guess `encodeTransaction` or `aminoWrap` throws some exceptions

**Comment by hewigovens** [comment]
BTW, can you run `clang-format` for the `Binance/SignerTests.cpp`?

File: tests/Binance/SignerTests.cpp [hunk 1]
```diff
@@ -202,4 +202,122 @@ TEST(BinanceSigner, BuildSend2) {
...
+        const auto fromWallet = HDWallet("patrol tired convince excite giant list tissue spring venture between rubber child kite hurry bulk tired raccoon beyond wrist naive message buzz ice slice", "");
+        const auto fromPrivateKey = fromWallet.getKey(derivationPath);
+        const auto fromPublicKey = PublicKey(fromPrivateKey.getPublicKey(TWPublicKeyTypeSECP256k1));
+        auto fromAddr = Cosmos::Address(HRP_BINANCE, fromPublicKey).keyHash;
+
+        const auto toWallet = HDWallet( "recycle poet history join say present until laptop normal burger inject entire end beach talent occur depth ask wage purpose tornado hold easily kick", "");
+        const auto toPrivateKey = toWallet.getKey(derivationPath);
+        const auto toPublicKey = PublicKey(toPrivateKey.getPublicKey(TWPublicKeyTypeSECP256k1));
+        auto toAddr = Cosmos::Address(HRP_BINANCE, toPublicKey).keyHash;
+
```
Comment (vikmeup) [inline tests/Binance/SignerTests.cpp]: can you use one private key for all examples? no need to convert from mnemonic => private key

File: src/proto/Binance.proto [hunk 2]
```diff
@@ -75,6 +75,40 @@ message TokenUnfreezeOrder {
     int64 amount = 3;  // amount of token to unfreeze
 }
 
+message HTLTOrder {
+    // 0xB33F9A24      // prefix
+    bytes from = 1;    // signer address
+    bytes to = 2;    // recipient address
+    string recipientOtherChain = 3;
+    string senderOtherChain = 4;
+    bytes randomNumberHash = 5;
```
Comment (vikmeup) [inline src/proto/Binance.proto]: can you use snake case as suggested by guidelines: https://developers.google.com/protocol-buffers/docs/style ?

File: src/proto/Binance.proto [hunk 3]
```diff
@@ -75,6 +75,40 @@ message TokenUnfreezeOrder {
...
+    int64 timestamp = 6;
+    repeated SendOrder.Token amount = 7;
+    string expectedIncome = 8;
+    int64 heightSpan = 9;
+    bool crossChain = 10;
+}
+
+message DepositHTLTOrder {
+    // 0xB33F9A24      // prefix
```
Comment (vikmeup) [inline src/proto/Binance.proto]: what is this prefix used for?

Comment (HaoyangLiu) [reply src/proto/Binance.proto]: https://github.com/trustwallet/wallet-core/pull/627/files#diff-9b3227fc67c8cc727f5c754a916eac8fR29
It is just a comment

**Comment by HaoyangLiu** [comment]
@hewigovens 
I have solve the issue. Now the new four transaction for Binance Chain are ready.

File: tests/Binance/SignerTests.cpp [hunk 4]
```diff
@@ -4,202 +4,385 @@
...
+        order.set_side(1);
+        order.set_price(100000000);
+        order.set_quantity(1200000000);
+        order.set_timeinforce(1);
+
+        auto signer = Binance::Signer(std::move(input));
+        auto result = signer.build();
+
+        ASSERT_EQ(
+                hex(result.begin(), result.end()),
```
Comment (vikmeup) [inline tests/Binance/SignerTests.cpp]: can you use expected value as 1 hex string?

File: tests/Binance/SignerTests.cpp [hunk 5]
```diff
@@ -4,202 +4,385 @@
...
+
+        ASSERT_EQ(hex(result.begin(), result.end()),
+                  "cc01"
+                  "f0625dee"
+                  "0a4e"
+                  "2a2c87fa"
+                  "0a23"
+                  "0a1440c2979694bbc961023d1d27be6fc4d21a9febe6120b0a03424e4210c098a8dd03"
+                  "1223"
+                  "0a1488b37d5e05f3699e2a1406468e5d87cb9dcceb95120b0a03424e4210c098a8dd03"
```
Comment (vikmeup) [inline tests/Binance/SignerTests.cpp]: same here

File: tests/Binance/SignerTests.cpp [hunk 6]
```diff
@@ -4,202 +4,385 @@
...
+                  "74657374"
+                  "2001");
+    }
+
+    TEST(BinanceSigner, BuildSend2) {
+        const auto derivationPath = TW::derivationPath(TWCoinTypeBinance);
+
+        const auto fromWallet =
+                HDWallet("swift slam quote sail high remain mandate sample now stamp title among fiscal "
+                         "captain joy puppy ghost arrow attract ozone situate install gain mean",
```
Comment (vikmeup) [inline tests/Binance/SignerTests.cpp]: use private key only , no need to use mnemonic phrase here

File: src/proto/Binance.proto [hunk 7]
```diff
@@ -75,6 +75,40 @@ message TokenUnfreezeOrder {
...
+message HTLTOrder {
+    // 0xB33F9A24      // prefix
+    bytes from = 1;    // signer address
+    bytes to = 2;    // recipient address
+    string recipient_other_chain = 3;
+    string sender_other_chain = 4;
+    bytes random_number_hash = 5;
+    int64 timestamp = 6;
+    repeated SendOrder.Token amount = 7;
+    string expected_income = 8;
```
Comment (vikmeup) [inline src/proto/Binance.proto]: `expected_income` what does this property mean?

File: src/Binance/Serialization.cpp [hunk 8]
```diff
@@ -19,6 +20,12 @@ static inline std::string addressString(const std::string& bytes) {
     return address.string();
 }
 
+static inline std::string encodeHexBytes(const std::string& bytes) {
```
Comment (vikmeup) [inline src/Binance/Serialization.cpp]: should we move this function into `HexCoding` ? @hewigovens

File: src/proto/Binance.proto (hunk 7 above)
Comment (HaoyangLiu) [reply src/proto/Binance.proto]: Please refer to the newest document: https://github.com/binance-chain/BEPs/blob/ea176af810196c21d6b5567ff7090c6247946ff0/BEP3.md

File: src/Binance/Serialization.cpp (hunk 8 above)
Comment (HaoyangLiu) [reply src/Binance/Serialization.cpp]: `encodeHexBytes` will generate uppercase string

File: src/proto/Binance.proto (hunk 7 above)
Comment (hewigovens) [reply src/proto/Binance.proto]: Can you add this as a comment? 

> expected gained token on the other chain

File: tests/Binance/SignerTests.cpp [hunk 9]
```diff
@@ -4,202 +4,312 @@
...
-            "1225""0a146b571fc0a9961a7ddf45e49a88a4d83941fcabbe120d0a03424e42108080e983b1de16"
-        "126c"
-            "0a26"
-            "eb5ae987"
-            "21027e69d96640300433654e016d218a8d7ffed751023d8efe81e55dedbd6754c971"
-            "1240""8b23eecfa8237a27676725173e58154e6c204bb291b31c3b7b507c8f04e2773909ba70e01b54f4bd0bc76669f5712a5a66b9508acdf3aa5e4fde75fbe57622a1"
-            "2001"
-    );
-}
+    TEST(BinanceSigner, Sign) {
```
Comment (hewigovens) [inline tests/Binance/SignerTests.cpp]: can you keep the old indent? otherwise too many unrelated changes

File: tests/Binance/SignerTests.cpp (hunk 4 above)
Comment (hewigovens) [reply tests/Binance/SignerTests.cpp]: +1, easier to copy

File: src/Binance/Serialization.cpp (hunk 8 above)
Comment (hewigovens) [reply src/Binance/Serialization.cpp]: must be upper case?

Comment (HaoyangLiu) [reply src/Binance/Serialization.cpp]: Previously it was. Now we have changed our chain. So lowercase is ok

Comment (hewigovens) [reply src/Binance/Serialization.cpp]: Do we still need this then?

Comment (HaoyangLiu) [reply src/Binance/Serialization.cpp]: Deleted
